### PR TITLE
fix deadlock

### LINKED ISSUE #2595: Synchronizer L1 parallel: get stuck after a new ForkId
# Rationale
Fix deadlock

# Implementation

Do you have ideas regarding the implementation of this feature?
Are you willing to implement this feature?

### PR DESCRIPTION
**Pull request by joanestebanr**
Closes #2595 

### What does this PR do?
Fix a problem of deadlock when a forkid produce a reset of state

### Reviewers
- @ARR552

### CODE REVIEW

File: synchronizer/ext_control.go [hunk 1]
```diff
@@ -0,0 +1,112 @@
+package synchronizer
```
Comment (ARR552) [inline synchronizer/ext_control.go:1]: Can you add somewhere a readme with an example of the file content to send commands?

File: synchronizer/ext_control.go [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+	"os"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/0xPolygonHermez/zkevm-node/log"
+)
+
+const (
+	externalControlFilename = "/tmp/synchronizer_in"
```
Comment (ARR552) [inline synchronizer/ext_control.go:15]: if this path must be created and filled with the commands, why don't we convert it in a config param?
I understand that if we runt the docker, first we should copy the file into the container. Am I right?

File: synchronizer/l1_rollup_info_consumer.go [hunk 3]
```diff
@@ -14,7 +14,7 @@ import (
 )
 
 const (
-	minNumIterationsBeforeStartCheckingTimeWaitinfForNewRollupInfoData = 20
+	minNumIterationsBeforeStartCheckingTimeWaitinfForNewRollupInfoData = 5
```
Comment (ARR552) [inline synchronizer/l1_rollup_info_consumer.go]:
```suggestion
	minNumIterationsBeforeStartCheckingTimeWaitingForNewRollupInfoData = 5
```

File: synchronizer/l1_rollup_info_consumer.go [hunk 4]
```diff
@@ -92,7 +92,7 @@ func (l *l1RollupInfoConsumer) step() error {
 		}
 		if rollupInfo.ctrlIsValid {
 			err = l.processIncommingRollupControlData(rollupInfo.ctrl)
-			if err != nil && !errors.Is(err, errConsumerStoppedBecauseIsSynchronized) {
+			if err != nil && !errors.Is(err, errConsumerStoppedBecauseIsSynchronized) && !errors.Is(err, errConsumerStopped) {
 				log.Error("consumer: error processing package.ControlData. Error: ", err)
```
Comment (ARR552) [inline synchronizer/l1_rollup_info_consumer.go:96]: In this case what happens?? would it retries to get the package in the next iteration?

File: synchronizer/l1_rollup_info_producer.go [hunk 5]
```diff
@@ -164,13 +164,14 @@ func newL1DataRetriever(cfg configProducer, ethermans []EthermanInterface, outgo
 // ResetAndStop: reset the object and stop the current process. Set first block to be retrieved
 func (l *l1RollupInfoProducer) ResetAndStop(startingBlockNumber uint64) {
 	log.Infof("producer: Reset L1 sync process to blockNumber %d st=%s", startingBlockNumber, l.toStringBrief())
-	l.mutex.Lock()
-	defer l.mutex.Unlock()
 	log.Debugf("producer: Reset(%d): stop previous run (state=%s)", startingBlockNumber, l.status.String())
-	l.stopUnsafe()
+	l.Stop()
```
Comment (ARR552) [inline synchronizer/l1_rollup_info_producer.go:168]: Is the stop called from somewhere without needing the lock?? Could we just have the stop() function, instead the stop() and stopUnsafe()???

File: synchronizer/l1_sync_orchestration.go [hunk 6]
```diff
@@ -58,20 +60,15 @@ func newL1SyncOrchestration(ctx context.Context, producer l1RollupProducerInterf
 
 func (l *l1SyncOrchestration) reset(startingBlockNumber uint64) {
 	log.Warnf("Reset L1 sync process to blockNumber %d", startingBlockNumber)
-	l.mutex.Lock()
-	defer l.mutex.Unlock()
-	comuserWasRunning := l.consumerStarted
-	if comuserWasRunning {
+	if l.isStarted {
```
Comment (ARR552) [inline synchronizer/l1_sync_orchestration.go]:
```suggestion
	if l.isRunning {
```
Just to use the same nomenclature than in line 53 and 54.

File: synchronizer/l1_sync_orchestration_test.go [hunk 7]
```diff
@@ -31,8 +31,8 @@ func TestGivenOrquestrationWhenHappyPathThenReturnsBlockAndNoErrorAndProducerIsR
 	returnedBlock, err := sut.start()
 	require.NoError(t, err)
 	require.Equal(t, block, *returnedBlock)
-	require.Equal(t, true, sut.producerStarted)
```
Comment (ARR552) [inline synchronizer/l1_sync_orchestration_test.go:34]: Why sut?? Shouldn't it be called sot??? (**S**etup**O**rchestration**T**est)

**Review by ARR552 — CHANGES_REQUESTED** [review]

File: synchronizer/ext_control.go (hunk 2 above)
Comment (joanestebanr) [reply synchronizer/ext_control.go:15]: in fact no, the code create the file and if exists skip all contents (is for live commands). I think that is not worth to allow configure it becuase is for debug/test propouses. If you want to use inside a docker you can map the volume of /tmp/ or the file by itself

File: synchronizer/l1_sync_orchestration_test.go (hunk 7 above)
Comment (joanestebanr) [reply synchronizer/l1_sync_orchestration_test.go:34]: sut stand for Subject Under Test... is a common way to name the object that you are testing

File: synchronizer/l1_sync_orchestration.go (hunk 6 above)
Comment (joanestebanr) [reply synchronizer/l1_sync_orchestration.go]: good point, done

File: synchronizer/l1_rollup_info_producer.go (hunk 5 above)
Comment (joanestebanr) [reply synchronizer/l1_rollup_info_producer.go:168]: Yes, the function stopUnsafe is already deleted!

File: synchronizer/l1_rollup_info_consumer.go (hunk 4 above)
Comment (joanestebanr) [reply synchronizer/l1_rollup_info_consumer.go:96]: It returns to main Sync function and depending of the lastblock returned is going to reset the synchronization process. This if is just to log an error or not. If the consumer finish by a controled situation we dont log a error

**Review by ARR552 — APPROVED** [review]

**Pull request merged**